File: src/tipos/pgcrp.ts
// Tipos para el sistema PGCRP (Actividades Extracátedra)

export interface Pgcrp {
  id_pgcrp: number;
  nombre: string;
  descripcion?: string;
  activo: boolean;
  fecha_creacion?: string;
}

export interface PgcrpAsignacionSeccion {
  id_asignacion_seccion: number;
  id_grado_secciones: number;
  id_pgcrp: number;
  id_periodo: number;
  fecha_asignacion?: string;
  activo: boolean;
  // Campos adicionales para consultas
  nombre_pgcrp?: string;
  grado?: string;
  seccion?: string;
  modalidad?: string;
}

export interface PgcrpAsignacionEstudiante {
  id_asignacion_estudiante: number;
  id_estudiante: number;
  id_pgcrp: number;
  id_periodo: number;
  fecha_asignacion?: string;
  activo: boolean;
  observaciones?: string;
  // Campos adicionales para consultas
  nombre_pgcrp?: string;
  nombres_estudiante?: string;
  apellidos_estudiante?: string;
  cedula_estudiante?: number;
}

export interface PgcrpEstudianteVista {
  id_estudiante: number;
  cedula: number;
  nombres: string;
  apellidos: string;
  id_grado_secciones: number;
  id_periodo: number;
  id_pgcrp_asignado?: number;
  nombre_pgcrp_asignado?: string;
  tipo_asignacion: 'individual' | 'seccion' | 'sin_asignar';
  observaciones_individuales?: string;
}

export interface AsignacionSeccionInput {
  id_grado_secciones: number;
  id_pgcrp: number;
  id_periodo: number;
}

export interface AsignacionEstudianteInput {
  id_estudiante: number;
  id_pgcrp: number;
  id_periodo: number;
  observaciones?: string;
}

export interface PgcrpInput {
  nombre: string;
  descripcion?: string;
}

// Tipos para reportes y estadísticas
export interface PgcrpEstadisticas {
  total_estudiantes: number;
  estudiantes_con_asignacion: number;
  estudiantes_sin_asignacion: number;
  asignaciones_por_seccion: number;
  asignaciones_individuales: number;
  actividades_utilizadas: ActividadEstadistica[];
}

export interface ActividadEstadistica {
  id_pgcrp: number;
  nombre_actividad: string;
  total_estudiantes: number;
  asignaciones_seccion: number;
  asignaciones_individuales: number;
}

export interface ReportePgcrpEstudiante {
  cedula: number;
  nombres: string;
  apellidos: string;
  grado: string;
  seccion: string;
  actividad_pgcrp: string;
  tipo_asignacion: string;
  observaciones?: string;
}

// Tipos para filtros y selecciones
export interface FiltrosPgcrp {
  id_periodo?: number;
  id_grado?: number;
  id_modalidad?: number;
  id_grado_secciones?: number;
}

// Tipos para la interfaz de gestión
export interface EstadoGestionPgcrp {
  actividades: Pgcrp[];
  asignacionesSeccion: PgcrpAsignacionSeccion[];
  asignacionesEstudiante: PgcrpAsignacionEstudiante[];
  estudiantesVista: PgcrpEstudianteVista[];
  estadisticas?: PgcrpEstadisticas;
  cargando: boolean;
  error?: string;
} 

File: src-tauri/src/models/pgcrp.rs
use serde::{Serialize, Deserialize};
use chrono::NaiveDateTime;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Pgcrp {
    pub id_pgcrp: i32,
    pub nombre: String,
    pub descripcion: Option<String>,
    pub activo: bool,
    pub fecha_creacion: Option<NaiveDateTime>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PgcrpAsignacionSeccion {
    pub id_asignacion_seccion: i32,
    pub id_grado_secciones: i32,
    pub id_pgcrp: i32,
    pub id_periodo: i32,
    pub fecha_asignacion: Option<NaiveDateTime>,
    pub activo: bool,
    // Campos adicionales para consultas
    pub nombre_pgcrp: Option<String>,
    pub grado: Option<String>,
    pub seccion: Option<String>,
    pub modalidad: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PgcrpAsignacionEstudiante {
    pub id_asignacion_estudiante: i32,
    pub id_estudiante: i32,
    pub id_pgcrp: i32,
    pub id_periodo: i32,
    pub fecha_asignacion: Option<NaiveDateTime>,
    pub activo: bool,
    pub observaciones: Option<String>,
    // Campos adicionales para consultas
    pub nombre_pgcrp: Option<String>,
    pub nombres_estudiante: Option<String>,
    pub apellidos_estudiante: Option<String>,
    pub cedula_estudiante: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PgcrpEstudianteVista {
    pub id_estudiante: i32,
    pub cedula: i64,
    pub nombres: String,
    pub apellidos: String,
    pub id_grado_secciones: i32,
    pub id_periodo: i32,
    pub id_pgcrp_asignado: Option<i32>,
    pub nombre_pgcrp_asignado: Option<String>,
    pub tipo_asignacion: String, // 'individual', 'seccion', 'sin_asignar'
    pub observaciones_individuales: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AsignacionSeccionInput {
    pub id_grado_secciones: i32,
    pub id_pgcrp: i32,
    pub id_periodo: i32,
}

#[derive(Debug, Deserialize)]
pub struct AsignacionEstudianteInput {
    pub id_estudiante: i32,
    pub id_pgcrp: i32,
    pub id_periodo: i32,
    pub observaciones: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct PgcrpInput {
    pub nombre: String,
    pub descripcion: Option<String>,
}

// Estructura para reportes y estadísticas
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PgcrpEstadisticas {
    pub total_estudiantes: i32,
    pub estudiantes_con_asignacion: i32,
    pub estudiantes_sin_asignacion: i32,
    pub asignaciones_por_seccion: i32,
    pub asignaciones_individuales: i32,
    pub actividades_utilizadas: Vec<ActividadEstadistica>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ActividadEstadistica {
    pub id_pgcrp: i32,
    pub nombre_actividad: String,
    pub total_estudiantes: i32,
    pub asignaciones_seccion: i32,
    pub asignaciones_individuales: i32,
}

// Estructura para el reporte de actividades extracátedra
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ReportePgcrpEstudiante {
    pub cedula: i64,
    pub nombres: String,
    pub apellidos: String,
    pub grado: String,
    pub seccion: String,
    pub actividad_pgcrp: String,
    pub tipo_asignacion: String,
    pub observaciones: Option<String>,
} 